qualtran/bloqs/gf_arithmetic/gf2_multiplication.py

```diff
@@ -653,3 +660,333 @@ def _binarypolynomialmultiplication() -> BinaryPolynomialMultiplication:
+            f=f_arr[: self.k],
+            g=g_arr[: self.k],
+            h=h_arr[: 2 * self.k - 1],
+        )
+
+        for i in range(self.n - self.k):
+            g_arr[self.k + i], g_arr[i] = bb.add(CNOT(), ctrl=g_arr[self.k + i], target=g_arr[i])
+        for i in range(self.n - self.k):
+            f_arr[self.k + i], f_arr[i] = bb.add(CNOT(), ctrl=f_arr[self.k + i], target=f_arr[i])
+
+        h = bb.join(h_arr[: self.n][::-1], self.qgf)
+        h = bb.add(
+            GF2MultiplyByConstantMod.from_polynomials(
+                [0, self.k], self.m_x, field_representation='int'
+            ).adjoint(),
+            g=h,
+        )
+        h_arr[: self.n] = bb.split(h)[::-1]
+        f_arr[self.k : self.n], g_arr[self.k : self.n], h_arr[: 2 * (self.n - self.k) - 1] = bb.add(
+            BinaryPolynomialMultiplication(self.n - self.k),
+            f=f_arr[self.k : self.n],
+            g=g_arr[self.k : self.n],
+            h=h_arr[: 2 * (self.n - self.k) - 1],
+        )
+
+        h = bb.join(h_arr[: self.n][::-1], self.qgf)
+        h = bb.add(GF2ShiftRightMod(self.m_x, self.k), f=h)
+        h_arr[: self.n] = bb.split(h)[::-1]
+
+        f_arr[: self.k], g_arr[: self.k], h_arr[: 2 * self.k - 1] = bb.add(
+            BinaryPolynomialMultiplication(self.k),
+            f=f_arr[: self.k],
+            g=g_arr[: self.k],
+            h=h_arr[: 2 * self.k - 1],
+        )
+        h = bb.join(h_arr[: self.n][::-1], self.qgf)
+        h = bb.add(
+            GF2MultiplyByConstantMod.from_polynomials(
+                [0, self.k], self.m_x, field_representation='int'
+            ),
+            g=h,
+        )
+        h_arr[: self.n] = bb.split(h)[::-1]
+
+        f_arr = f_arr[::-1]
+        g_arr = g_arr[::-1]
+        h_arr = h_arr[::-1]
+        f = bb.join(f_arr, self.qgf)
+        g = bb.join(g_arr, self.qgf)
+        h = bb.join(h_arr, self.qgf)
+        return {'f': f, 'g': g, 'h': h}
+
+
+@attrs.frozen
+class GF2Mul(Bloq):
+    r"""Multiplies two GF($2^n$) numbers (or binary polynomials) modulu.
+
+    Applies the transformation
+    $$
+        \ket{f}\ket{g} \rightarrow \ket{f} \ket{g} \ket{f*g \mod m(x)}
+    $$
+
+    Where the modulus $m(x)$ is the irreducible polynomial defining the galois field arithmetic.
+    The toffoli complexity is $n^{\log_2{3}}$
+
+    Args:
+        m_x: The irreducible polynomial that defines the galois field.
+        uncompute: Whether to compute or uncompute the product.
+
+    Registers:
+        f: The first number (polynomial).
+        g: The second number (polynomial).
+        h: The result.
+
+    References:
+        [Space-efficient quantum multiplication of polynomials for binary finite fields with
+            sub-quadratic Toffoli gate count](https://arxiv.org/abs/1910.02849v2) Algorithm 4.
+    """
+
+    m_x: Poly = attrs.field(converter=lambda x: x if isinstance(x, Poly) else Poly.Degrees(x))
+    uncompute: bool = False
```

Comment:
Please change signature to `dtype: QGF` and `uncompute: bool`. It would also be nice if you can support a `plus_equal_prod: bool = False` parameter similar to `GF2Multiplication` bloq, but that can also be left for future improvement.

Reply:
done